cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.planner;
+
+import co.cask.cdap.etl.api.Transform;
+import co.cask.cdap.etl.common.PipelinePhase;
+import co.cask.cdap.etl.proto.Connection;
+import co.cask.cdap.etl.spec.PipelineSpec;
+import co.cask.cdap.etl.spec.StageSpec;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+
+/**
+ * Takes a {@link PipelineSpec} and creates an execution plan from it.
+ */
+public class PipelinePlanner {
+  private final String sourceType;
+  private final String sinkType;
+  private final Set<String> reduceTypes;
+
+  public PipelinePlanner(String sourceType, String sinkType, Set<String> reduceTypes) {
+    this.sourceType = sourceType;
+    this.sinkType = sinkType;
+    this.reduceTypes = ImmutableSet.copyOf(reduceTypes);
+  }
+
+  /**
+   * Create an execution plan for the given logical pipeline.
+   *
+   * A plan consists of one or more phases, with connections between phases.
+   * A connection between a phase indicates control flow, and not necessarily
+   * data flow. This class assumes that it receives a valid pipeline spec.
+   * That is, the pipeline has no cycles, all its nodes have unique names,
+   * sources don't have any input, sinks don't have any output,
+   * everything else has both an input and an output, etc.
+   *
+   * We start by inserting connector nodes into the logical dag,
+   * which are used to mark boundaries between mapreduce jobs.
+   * Each connector represents a node where we will need to write to a local dataset.
```

Comment:
yep the most direct use case is for batch pipelines and using local datasets.  Though for realtime pipelines, it may still be useful to mark as a point where we need to do windowing?  Would need to think through it more though.